src/plugins/elasticsearch_ui_shared/public/request.ts

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { useEffect, useState, useRef } from 'react';
+
+export interface SendRequest {
+  path: string;
+  method: string;
+  body?: any;
+  uimActionType?: string;
+}
+
+export interface SendRequestResponse {
+  data: any;
+  error: Error;
+}
+
+export interface UseRequest extends SendRequest {
+  interval?: number;
+  initialData?: any;
+  processData?: any;
+}
+
+export const sendRequest = async (
+  httpClient: any,
+  { path, method, body }: SendRequest
+): Promise<Partial<SendRequestResponse>> => {
+  try {
+    // NOTE: This is tightly coupled to Angular's $http service.
+    const response = await httpClient[method](path, body);
+
+    if (typeof response.data === 'undefined') {
+      throw new Error(response.statusText);
+    }
+
+    return { data: response.data };
+  } catch (e) {
+    return {
+      error: e.response ? e.response : e,
+    };
+  }
+};
+
+export const useRequest = (
+  httpClient: any,
+  { path, method, body, interval, initialData, processData }: UseRequest
```

Comment:
`automaticallyResendRequestEvery` is too verbose imo. some alternatives: `requestIntervalMs`, `pollingIntervalMs`

Reply:
I agree that `automaticallyResendRequestEvery` is not the best, was a quick attempt. But all the `interval`, `requestIntervalMs`, `pollingIntervalMs` imply that we somehow now that there is a mechanism to resend the request. Which is, actually, an edge case that we use for our list tables.

What if we had an option object `scheduleRequest` with the following shape?

```ts
{ enabled: true, interval: 5000 };
```

Reply:
I think `pollIntervalMs` is a good combination of terseness and clarity.

>  imply that we somehow now that there is a mechanism to resend the request. 

I'm not sure I follow. `useRequest` implements logic to resend the request if the interval is specified, on lines 88 and 124. Once a request resolves, it schedules the next request. What am I missing?

> Which is, actually, an edge case that we use for our list tables.

IMO the ability to set a polling interval is one of the really powerful and useful features of this hook, since any UX that depends upon data we expect will change often can use it. One example of polling outside of tables is Console, which polls for changes to the fields in a user's data, so that the autocomplete feature can suggest them to the user.

It seems like specifying an interval or not is sufficient to infer whether the consumer wants to poll or not -- could you explain the benefit of the object you're proposing?

> Also, what do you think to set a default "processData" or "serializer" method?

Sure, I'll make that change to make it more obvious this parameter is optional.

Reply:
I am not arguing that having a poll mechanism is awesome. But it is only useful for the `GET` method in a few places. I see `useRequest` as a hooked **http client** to make  HTTP requests in a component and get a Response from the server. We will never poll `POST`, `PUT`, `DELETE`, `PATCH` methods.

My point was, as a consumer, this is a cool extra feature. But this is not expected from an http client, so I thought it was better to make it explicit that we can schedule requests. (it is on lines 88 and 124 but the consumer does not know about that).

Just sharing my first experience with the API.